Add support for 'raj' locale in the application

rahul-vyas-dev opened the pull request:
## Description

This pull request adds support for the Rajasthani language to the project.
The update includes new translation files, proper localization setup, and integration with the existing i18n workflow.

This change ensures users can switch to the Rajasthani language and experience the UI fully translated.

Fixes #604 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Tested the Rajasthani translation file locally by running the project.

- Verified that all keys display correctly and no missing translation warnings appear.

- Manually checked each updated screen to confirm the translated strings are rendered as expected.

- Tested switching languages from the settings menu to ensure Rajasthani is selectable.

Test Steps:

- Pull the latest codebase.

- Switch the app language to Rajasthani.

- Navigate through pages to confirm translations load correctly.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

## Maintainer Checklist

- [x] closes #604 
- [ ] Tag the PR with the appropriate labels

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Rajasthani (raj) language added across the app with full translations, pluralization, and dynamic messages.
  * Locale selection and runtime localization updated so UI, notifications, prompts and accessibility labels support raj.

* **Chores**
  * Added Raj locale resources and date/time formatting data to ensure correct display and locale-aware formatting.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rahul-vyas-dev committed:
Add support for 'raj' locale in the application

rahul-vyas-dev committed:
Merge branch 'master' into feat/add-rajasthani-lang

M4dhav reviewed (changes requested):
Please provide a video to verify that this works in-app

rahul-vyas-dev commented:
Hey @M4dhav  ! Actually my Flutter setup got corrupted recently, and I’m facing issues getting the environment working again (I also mentioned this in the Discord server). Because of that, I’m not able to re-run the project right now, and I didn’t record any video when I initially tested it.

rahul-vyas-dev commented:
@M4dhav if you want any changes then let me know i am here for that

rahul-vyas-dev commented:
hey @M4dhav I verify the ARB messages using the check_arb.dart script, and the output looks correct. Here’s what I got:

dart "c:\Users\rahul\GSOC\Resonate\tools\check_arb.dart"
Loaded ARB: lib/l10n/app_raj.arb
Locale: raj
Messages found: 362

 - title => रेज़ोनेट
 - noAvailableRoom =>  false{कोई आवण वालो रूम नी है} other{रूम री जानकारी नी है}}
नीचे एक जोड़ण सूं शुरू करो!
 - noAvailableRoom#2 =>  false{कोई आवण वालो रूम नी है} other{रूम री जानकारी नी है}}
नीचे एक जोड़ण सूं शुरू करो!
 - minutesAgo =>  other{1 मिनट पहलां}}
 - minutesAgo#2 =>  other{5 मिनट पहलां}}

# Code that i have used

```dart
import 'dart:convert';
import 'dart:io';

void main(List<String> args) async {
  final path = args.isNotEmpty ? args.first : 'lib/l10n/app_raj.arb';
  final file = File(path);
  if (!await file.exists()) {
    stderr.writeln('File not found: $path');
    exit(2);
  }

  final content = await file.readAsString();
  Map<String, dynamic> json;
  try {
    json = jsonDecode(content) as Map<String, dynamic>;
  } catch (e) {
    stderr.writeln('Invalid JSON in $path: $e');
    exit(3);
  }

  stdout.writeln('Loaded ARB: $path');
  stdout.writeln('Locale: ${json['@@locale'] ?? '<none>'}');
  stdout.writeln(
    'Messages found: ${json.keys.where((k) => !k.startsWith('@')).length}',
  );
  stdout.writeln('');

  final tests = <String, Map<String, Object?>>{
    'title': {},
    'noAvailableRoom': {'isRoom': 'true'},
    'noAvailableRoom#2': {'isRoom': 'false'},
    'minutesAgo': {'count': 1},
    'minutesAgo#2': {'count': 5},
  };

  for (final entry in tests.entries) {
    final rawKey = entry.key.contains('#')
        ? entry.key.split('#').first
        : entry.key;
    final sampleName = entry.key;
    final argsForKey = entry.value;

    final raw = json[rawKey];
    if (raw == null) {
      stdout.writeln(' - $sampleName: <missing key in ARB>');
      continue;
    }

    final formatted = formatMessage(raw.toString(), argsForKey);
    stdout.writeln(' - $sampleName => $formatted');
  }
}

String formatMessage(String msg, Map<String, Object?> args) {
  var out = msg;

  final selectRe = RegExp(r'\{(\w+),\s*select,([\s\S]*?)\}');
  out = out.replaceAllMapped(selectRe, (m) {
    final varName = m[1]!;
    final body = m[2]!;
    final choice = _pickSelectChoice(body, args[varName]?.toString());
    return choice;
  });

  final pluralRe = RegExp(r'\{(\w+),\s*plural,([\s\S]*?)\}');
  out = out.replaceAllMapped(pluralRe, (m) {
    final varName = m[1]!;
    final body = m[2]!;
    final count = args[varName];
    final choice = _pickPluralChoice(body, count);
    return choice;
  });

  final placeholderRe = RegExp(r'\{(\w+)\}');
  out = out.replaceAllMapped(placeholderRe, (m) {
    final n = m[1]!;
    if (args.containsKey(n) && args[n] != null) return args[n].toString();
    return '{${n}}';
  });

  out = out.replaceAll(r'\n', '\n');
  return out;
}

String _pickSelectChoice(String body, String? value) {
  final tokenRe = RegExp(r'(\w+)\{([^}]*)\}');
  final matches = tokenRe.allMatches(body);
  String? other;
  for (final m in matches) {
    final k = m[1]!;
    final v = m[2]!;
    if (k == 'other') other = v;
    if (value != null && value == k) return v;
  }
  return other ?? '';
}

String _pickPluralChoice(String body, Object? countObj) {
  final count = (countObj is num) ? countObj.toInt() : null;
  final tokenRe = RegExp(r'(=?\d+|other)\{([^}]*)\}');
  final matches = tokenRe.allMatches(body);
  String? other;
  final map = <String, String>{};
  for (final m in matches) {
    final k = m[1]!;
    final v = m[2]!;
    map[k] = v;
    if (k == 'other') other = v;
  }
  if (count != null) {
    final eqKey = '=${count}';
    if (map.containsKey(eqKey))
      return map[eqKey]!.replaceAll('{count}', count.toString());
    if (count == 0 && map.containsKey('=0'))
      return map['=0']!.replaceAll('{count}', '0');
    if (count == 1 && map.containsKey('=1'))
      return map['=1']!.replaceAll('{count}', '1');
    return (other ?? '').replaceAll('{count}', count.toString());
  }
  return other ?? '';
}
```

rahul-vyas-dev commented:
> Hey @rahul-vyas-dev I just checked your PR Locally, and the app does not, in fact, run with the `raj` localization added. You can verify this by trying to open the App Preferences page when the localization is present. Essentially that means that Flutter does not have support for 3 letter language codes via flutter_localization.


working on it

rahul-vyas-dev committed:
feat: add locale resolution for 'raj' language support

M4dhav commented on lib/main.dart:
```diff
@@ -89,6 +89,18 @@ class MyApp extends StatelessWidget {
+            if (supportedLocale.languageCode == locale?.languageCode) {
+              return supportedLocale;
+            }
+          }
```
why was code added to main.dart?

rahul-vyas-dev replied:
Sorry for the confusion! The changes in main.dart were added because this branch includes the new Rajasthani (raj) language support. To enable that, the app needed an updated localeResolutionCallback and localization setup inside GetMaterialApp

rahul-vyas-dev committed:
fixing the code by following main branch

rahul-vyas-dev committed:
Add Rajasthani localization support

- Introduced a new localization file `raj_intl.dart` containing date patterns and symbols for the Rajasthani language.
- Implemented the `AppLocalizationsRaj` class extending `GlobalMaterialLocalizations` to provide localized strings and formats.
- Updated `main.dart` to include the Rajasthani localization delegate in the app's localization setup.
Added support for Rajesthani Language

rahul-vyas-dev committed:
feat: add Rajasthani localization support and update untranslated strings

rahul-vyas-dev committed:
fixing the date format according to RAJ lang

rahul-vyas-dev committed:
fixing the import statement

M4dhav commented:
Please change base branch of PR to Dev

rahul-vyas-dev commented:
hey @M4dhav review please

M4dhav commented:
Hey @rahul-vyas-dev do you still want to work on this?

rahul-vyas-dev commented:
> Hey @rahul-vyas-dev do you still want to work on this?

Yes, I’d be happy to continue working on this.

rahul-vyas-dev committed:
Merge branch 'dev' into feat/add-rajasthani-lang

rahul-vyas-dev commented:
@M4dhav please review this

M4dhav commented on lib/l10n/raj_intl.dart:
Add translations for all the strings in this file

rahul-vyas-dev replied:
@M4dhav

rahul-vyas-dev committed:
adding all untranslated messages for rajesthani

rahul-vyas-dev committed:
Merge branch 'feat/add-rajasthani-lang' of https://github.com/rahul-vyas-dev/Resonate into feat/add-rajasthani-lang

rahul-vyas-dev commented:
@M4dhav

rahul-vyas-dev committed:
Merge branch 'dev' into feat/add-rajasthani-lang
